RecoCTPPS/TotemRPLocal/interface/CTPPSTimingTrackRecognition.h

```diff
@@ -0,0 +1,305 @@
+     * @getHitCenter: function extracting hit's center in the dimension
+     *      that the partial tracks relate to
+     * @getHitRangeWidth: analogical to getHitCenter, but extracts hit's width
+     *      in specific dimension
+     * @setTrackCenter: function used to set track's position in considered dimension
+     * @setTrackSigma: function used to set track's sigma in considered dimension
+     * @result: vector to which produced tracks are appended
+     */
+    void producePartialTracks(
+        const HitVector& hits,
+        const DimensionParameters& param,
+        float (*getHitCenter)(const HIT_TYPE&),
+        float (*getHitRangeWidth)(const HIT_TYPE&),
+        void (*setTrackCenter)(TRACK_TYPE&, float),
+        void (*setTrackSigma)(TRACK_TYPE&, float),
+        std::vector<TRACK_TYPE>& result
+      );
+
+
+
+    /* Retrieves the bounds of a 3D range in which all hits from given collection are contained.
+     *
+     * @hits: hits collection to retrieve the range from
+     */
+    SpatialRange getHitSpatialRange(const HitVector& hits);
+};
+
+/****************************************************************************
+ * Implementation
+ ****************************************************************************/
+
+
+template<class TRACK_TYPE, class HIT_TYPE>
+CTPPSTimingTrackRecognition<TRACK_TYPE, HIT_TYPE>::CTPPSTimingTrackRecognition(const edm::ParameterSet& iConfig) :
+    threshold               ( iConfig.getParameter<double>( "threshold" ) ),
+    thresholdFromMaximum    ( iConfig.getParameter<double>( "thresholdFromMaximum" ) ),
+    resolution              ( iConfig.getParameter<double>( "resolution" ) ),
+    sigma                   ( iConfig.getParameter<double>( "sigma" ) ),
+    pixelEfficiencyFunction ( "hit_TF1_CTPPS", iConfig.getParameter<std::string>( "pixelEfficiencyFunction" ).c_str() ) {
+}
+
+
+template<class TRACK_TYPE, class HIT_TYPE>
+CTPPSTimingTrackRecognition<TRACK_TYPE, HIT_TYPE>::~CTPPSTimingTrackRecognition() {};
+
+
+template<class TRACK_TYPE, class HIT_TYPE>
+void CTPPSTimingTrackRecognition<TRACK_TYPE, HIT_TYPE>::clear() {
+  hitVectorMap.clear();
+}
+
+
+template<class TRACK_TYPE, class HIT_TYPE>
+void CTPPSTimingTrackRecognition<TRACK_TYPE, HIT_TYPE>::producePartialTracks(
+    const HitVector& hits,
+    const DimensionParameters& param,
+    float (*getHitCenter)(const HIT_TYPE&),
+    float (*getHitRangeWidth)(const HIT_TYPE&),
+    void (*setTrackCenter)(TRACK_TYPE&, float),
+    void (*setTrackSigma)(TRACK_TYPE&, float),
+    std::vector<TRACK_TYPE>& result
+  ) {
+
+  int numberOfTracks = 0;
+  const double invResolution = 1./param.resolution;
+  const float profileRangeMargin = param.sigma * 3.;
+  const float profileRangeBegin = param.rangeBegin - profileRangeMargin;
+  const float profileRangeEnd = param.rangeEnd + profileRangeMargin;
+
+  std::vector<float> hitProfile((profileRangeEnd - profileRangeBegin) * invResolution, 0.);
+  auto hitFunction = param.hitFunction;
```

Comment:
Thanks for the suggestion! I am however wondering if such an optimisation is needed for our low-intensity TOTEM runs where the expected # digis / BX is rather low?

Reply:
Unfortunately it seems that `FormulaEvaluator` has a limited grammar with respect to `TF1` parsing capabilities. I.e. our new default efficiency function including a mixing of step functions cannot be evaluated:
```sh
----- Begin Fatal Exception 07-Jan-2019 16:07:22 CET-----------------------
An exception of category 'FormulaEvaluatorParseError' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=TotemTimingRecHitProducer label='totemTimingRecHits'
Exception Message:
While parsing '([0] + [1]*x + [2]*x^2)*(x<225) + ([0] + [1]*225 + [2]*225^2 + [3]*(x-225) + [4]*(x-225)^2 + [5]*(x-225)^3)*(x>225)' could not parse beyond '([0]'
----- End Fatal Exception -------------------------------------------------
```

Reply:
Oh, now I realise this comes from the other use of a `TF1` in our PR. This might then work if I revert back to `TF1` for the timing calibration helper object...

Reply:
@Dr15Jones  please check the issue above https://github.com/cms-sw/cmssw/pull/25465#discussion_r245686644 related to parsing 
```While parsing '([0] + [1]*x + [2]*x^2)*(x<225) + ([0] + [1]*225 + [2]*225^2 + [3]*(x-225) + [4]*(x-225)^2 + [5]*(x-225)^3)*(x>225)' could not parse beyond '([0]'```
Is there an alternative syntax or should we have an update to the FormulaEvaluator?

Reply:
It looks like the `FormulaEvaluator` does not like white space, try
```
([0]+[1]*x+[2]*x^2)*(x<225)+([0]+[1]*225+[2]*225^2+[3]*(x-225)+[4]*(x-225)^2+[5]*(x-225)^3)*(x>225)
```